src/db/gc/mod.rs

```diff
@@ -0,0 +1,323 @@
+//!
+//! ## Design goals
+//! A correct GC algorithm that is simple and efficient for forest scenarios. This algorithm removes
+//! unreachable blocks that are older than `chain finality`, making sure to avoid removing something
+//! that could later become reachable as a result of a fork.
+//!
+//! ## GC Algorithm
+//! The `mark-and-sweep` algorithm was chosen due to it's simplicity, efficiency and low memory
+//! footprint. Previously the `semi-space` algorithm was used resulting in data duplication and up
+//! to a 100% extra disk usage.
+//!
+//! ## GC Workflow
+//! 1. Mark: traverse all the relevant database columns, generating integer hash representations for
+//! each database key and storing those in a set.
+//! 2. Wait at least `chain finality` blocks.
+//! 3. Traverse reachable blocks starting at the current heaviest tipset and remove those from the
+//! marked set, leaving only unreachable entries that are older than `chain finality`.
+//! 4. Sweep, removing all the remaining marked entries from the database.
+//!
+//! ## Correctness
+//! This algorithm considers all the blocks that are visited during the `snapshot export` task
+//! reachable, making sure they are kept in the database after the run. It makes sure to retain the
+//! reachable graph as well as all the blocks for at least `chain finality` to account for potential
+//! forks. A snapshot can be used to bootstrap the node from scratch, thus the algorithm is
+//! considered correct when a valid snapshot can be exported using records available in the database
+//! after the run.
+//!
+//! ## Disk usage
+//! There's no additional disk space required to run this algorithm. However, removing the
+//! unreachable blocks from the database takes at least `chain finality`. The GC speed depends on
+//! the reachable graph size. Disk usage is also affected by the GC run interval. Additionally,
+//! since a truncated 4-byte hash is used - there's a slight possibility of a collision, which might
+//! result in an unreachable block being retained in the database. Still, the impact on the total
+//! disk size is negligible.
+//!
+//! ## Memory usage
+//! During the `mark` and up to the `sweep` stage, the algorithm requires `4 bytes` of memory for
+//! each database record. Additionally, the seen cache while traversing the reachable graph
+//! executing the `filter` stage requires at least `32 bytes` of memory for each reachable block.
+//!
+//! ## Scheduling
+//! 1. GC is triggered automatically and there have to be at least `chain finality` epochs stored
+//! for the `mark` step.
+//! 2. The `filter` step is triggered after at least `chain finality` has passed since the `mark`
+//! step.
+//! 3. Then, the `sweep` step happens.
+//! 4. Finally, the algorithm waits for a configured amount of time to initiate the next run.
+//!
+//! ## Performance
+//! Note: This needs to be measured and potentially defined in terms of `snapshot export` or any
+//! other visible and comparable metric.
```

Comment:
What can we say about the performance without doing any measurements? What are your expectations? If a node runs for a month, how many CPU seconds do you estimate the GC will use?

Reply:
I wouldn't be able to speculate about this without actually measuring this. Also, it depends on the node specs.

Reply:
We've reasoned about the disk use and the memory usage. Why can't we reason about the CPU use? Is there nothing you can say about the time complexity of this algorithm?

Reply:
In order to reason about CPU use we have to define that for graph walk algorithm first and that would depend on whether it's parallel or sequential. The rest is O(n).

Reply:
I'll contrive something.

Reply:
> In order to reason about CPU use we have to define that for graph walk algorithm first and that would depend on whether it's parallel or sequential. The rest is O(n).

What's the time complexity of a depth-first graph traversal? And what's the time complexity of a parallel depth-first graph traversal?

Reply:
Both are O(V+E), where v is the number of vertices and e is the number of edges.